flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/BatchingStateChangeStore.java

```diff
@@ -136,21 +133,23 @@ private void drainAndSave() {
             tasks = new ArrayList<>(scheduled);
             scheduled.clear();
             scheduledSizeInBytes = 0;
-            scheduledFuture = null;
         }
         try {
-            if (error != null) {
-                tasks.forEach(task -> task.fail(error));
+            if (error.get() != null) {
+                tasks.forEach(task -> task.fail(error.get()));
                 return;
             }
-            retryingExecutor.execute(retryPolicy, () -> delegate.save(tasks));
+            delegate.save(tasks);
+
+            synchronized (scheduled) {
+                scheduleUploadIfNeeded();
+            }
```

Comment:
This won't be executed in case of exception (but I guess it will change if concurrency issues are addressed).

But more important, no future uploads will start because `this.error` is set later in `catch` block. I guess the assumption is that a single failure leads to the whole job failover?

But that shouldn't be the case, as tasks should tolerate checkpoint failures and it's the JM who decides how to handle checkpoint failures.

Reply:
I think you are right, this should be moved to finally block?

Reply:
Yes, I think this would solve the 1st issue.
WDYT about the 2nd, more important one?